src/ooni/dns_test_impl.hpp

```diff
@@ -27,45 +29,42 @@ class DNSTestImpl : public ooni::OoniTestImpl {
                std::string query_name, std::string nameserver,
                std::function<void(dns::Message)> cb) {
 
-        std::string nameserver_part;
+        int resolver_port;
+        std::string resolver_hostname, nameserver_part;
         std::stringstream nameserver_ss(nameserver);
+
         std::getline(nameserver_ss, nameserver_part, ':');
-        entry["resolver"].push_back(nameserver_part);
+        resolver_hostname = nameserver_part;
+
         std::getline(nameserver_ss, nameserver_part, ':');
-        entry["resolver"].push_back(nameserver_part);
+        resolver_port = std::stoi(nameserver_part, nullptr);
 
         dns::query(
             query_class, query_type, query_name,
             [=](Error error, dns::Message message) {
                 logger.debug("dns_test: got response!");
-                YAML::Node query_entry;
+                json query_entry;
+                query_entry["resolver_hostname"] = resolver_hostname;
+                query_entry["resolver_port"] = resolver_port;
+                query_entry["failure"] = NULL;
```

Comment:
I recall that yaml-cpp converted `NULL` to zero rather than to `null`. I am not sure of what is done by this new json library. Could you check whether we end up with the desired value in the output? Otherwise, it's probably reasonable to expect that, to get `null` on output, one shall use `nullptr` in C++.